spark/src/main/scala/org/apache/spark/sql/delta/UpdateExpressionsSupport.scala

```diff
@@ -114,6 +131,25 @@ trait UpdateExpressionsSupport extends CastSupport with SQLConfHelper with Analy
                   ArrayType(toEt, containsNull = true)
                 )
             }
+          case NestedMapConversion(from: MapType, to: MapType) =>
+            if (!from.keyType.sameType(to.keyType)) {
+              throw DeltaErrors.unsupportedMapKeyTypeChange(
+                mapType = from,
+                fromKeyType = from.keyType,
+                toKeyType = to.keyType
+              )
+            }
+            val valueConverter: (Expression, Expression) => Expression = (key, _) =>
```

Comment:
We need to be more cautious about merge map struct evolution. This converts every element to the desired type, which will add some significant performance regression if the types are the same or castable.

https://databricks.atlassian.net/browse/ES-735586 is the ES ticket for the performance regression from Array of nested structs.

Reply:
Good call, I added a check to only convert when the types can't be directly cast.